src/vos/tests/vos_cmd.c

```diff
@@ -0,0 +1,1072 @@
+		}
+	}
+
+	D_ALLOC_PTR(known_pool);
+	if (known_pool == NULL)
+		return -DER_NOMEM;
+
+	D_STRNDUP(known_pool->kp_name, name, PATH_MAX);
+	if (known_pool->kp_name == NULL) {
+		D_FREE(known_pool);
+		return -DER_NOMEM;
+	}
+
+	create_const_uuid(known_pool->kp_uuid, name);
+
+	D_ASPRINTF(known_pool->kp_path, "%s/%s.vos", vos_path, name);
+	if (known_pool->kp_path == NULL) {
+		D_FREE(known_pool->kp_name);
+		D_FREE(known_pool);
+		return -DER_NOMEM;
+	}
+
+	d_list_add_tail(&known_pool->kp_link, &pool_list);
+	*pool = known_pool;
+	return 1;
+}
+
+static void
+free_pool(struct known_pool *pool)
+{
+	d_list_del(&pool->kp_link);
+	D_FREE(pool->kp_path);
+	D_FREE(pool->kp_name);
+	D_FREE(pool);
+}
+
+/* ACTION(name, function, open, randompct) */
+#define FOREACH_OP(ACTION)                                                                         \
+	ACTION(CREATE_POOL, create_pool, false, 0)                                                 \
+	ACTION(OPEN_POOL, open_pool, false, 0)                                                     \
+	ACTION(CLOSE_POOL, close_pool, true, 0)                                                    \
+	ACTION(PUNCH_KEY, punch_key, true, 2)                                                      \
+	ACTION(PUNCH_EXTENT, write_key, true, 18)                                                  \
+	ACTION(WRITE, write_key, true, 56)                                                         \
+	ACTION(REMOVE_ONE, write_key, true, 18)                                                    \
+	ACTION(REMOVE_ALL, write_key, true, 5)                                                     \
+	ACTION(AGGREGATE, aggregate, true, 1)                                                      \
+	ACTION(DISCARD, discard, true, 0)                                                          \
+	ACTION(ITERATE, iterate, true, 0)                                                          \
+	ACTION(SIZE_QUERY, print_size, true, 0)                                                    \
+	ACTION(RANDOMIZE, run_many_tests, true, 0)
+
+#define DEFINE_OP_ENUM(name, function, open, share) name,
+enum {
+	FOREACH_OP(DEFINE_OP_ENUM) OP_COUNT,
+};
+
+struct op_info {
+	const char *oi_str;
+	int (*oi_func)(struct cmd_info *);
+	bool oi_open;
+	int  oi_random_share;
+};
+
+static struct op_info op_info[];
+
+static int            shares[100];
+
+static void
+init_shares(void)
+{
+	int op;
+	int share;
+	int cursor = 0;
+	int c;
+
+	for (op = 0; op < OP_COUNT; op++)
+		for (share = 0; share < op_info[op].oi_random_share; share++)
+			shares[cursor++] = op;
+	/** The values defined in FOREACH_OP are percentages and should add up to 100 which should
+	 *  also be equal to the size of the array shares.should add up to the number of entries
+	 *  in the array.  If we change it to some other shares, they should still add up to the
+	 *  number in the array.
+	 */
+	D_ASSERT(cursor == ARRAY_SIZE(shares));
+
+	for (c = 0; c < sizeof(write_buf); c++)
+		write_buf[c] = (c % 26) + 97;
```

Comment:
My concern is that the for section will repeat for 16 * 1024 * 1024 times to initialize the write_buf. It is some inefficient.

Reply:
Yeah, it's a one time thing and takes 0.07 seconds on my 7 year old machine so not worth optimizing.